src/EditorFeatures/Core/IntelliSense/AsyncCompletion/CompletionSource.cs

```diff
@@ -282,103 +283,45 @@ internal sealed class CompletionSource : IAsyncExpandingCompletionSource
-                        // There could be a race around the usage of this stopwatch, I ignored it since we just need a rough idea:
-                        // we always log the time even if the stopwatch's not started regardless of whether expand items are included intially
-                        // (that number can be obtained via another property.)
-                        AsyncCompletionLogger.LogAdditionalTicksToCompleteDelayedImportCompletionDataPoint(stopwatch.Elapsed);
-
-                        return result;
-                    }, expandeditemsTaskCancellationToken);
+                    var expandedItemsTaskCancellationToken = _expandedItemsTaskCancellationSeries.CreateNext(cancellationToken);
+                    var expandedItemsTask = Task.Run(async () => await GetCompletionContextWorkerAsync(session, document, trigger, triggerLocation,
+                                                                        options with { ExpandedCompletionBehavior = ExpandedCompletionMode.ExpandedItemsOnly },
+                                                                        expandedItemsTaskCancellationToken).ConfigureAwait(false),
+                                                     expandedItemsTaskCancellationToken);
 
                     // Now trigger and wait for core providers to return;
                     var (nonExpandedContext, nonExpandedCompletionList) = await GetCompletionContextWorkerAsync(session, document, trigger, triggerLocation,
                             options with { ExpandedCompletionBehavior = ExpandedCompletionMode.NonExpandedItemsOnly }, cancellationToken).ConfigureAwait(false);
+
                     UpdateSessionData(session, sessionData, nonExpandedCompletionList, triggerLocation);
 
                     // If the core items are exclusive, we don't include expanded items.
+                    // This would cancel expandedItemsTask.
                     if (sessionData.IsExclusive)
                     {
-                        // This would cancel expandedItemsTask.
-                        _ = _expandeditemsTaskCancellationSeries.CreateNext(CancellationToken.None);
-                        return nonExpandedContext;
-                    }
-
-                    if (expandedItemsTask.IsCompleted)
-                    {
-                        // the task of expanded item is completed, get the result and combine it with result of non-expanded items.
-                        var (expandedContext, expandedCompletionList) = await expandedItemsTask.ConfigureAwait(false);
-                        UpdateSessionData(session, sessionData, expandedCompletionList, triggerLocation);
-                        AsyncCompletionLogger.LogImportCompletionGetContext(isBlocking: false, delayed: false);
-
-                        return CombineCompletionContext(session, nonExpandedContext, expandedContext);
+                        _ = _expandedItemsTaskCancellationSeries.CreateNext(CancellationToken.None);
```

Comment:
this needs docs as well.  i think i get it.  you're cancelling any outstanding work since it couldn't result in items that would matter (since you're in exclusive mode).  but needs docs to explain and be certain.

Reply:
question: if we're going to cancel the work anyways... can we not check if things are exclusive *first* before kicking off the work in the first place?

Reply:
Yes, your understanding is correct. I will doc this. The reason we kick off the "unimported" task first is because, waiting for the value of exclusivity from regular providers means computation of unimported items won't start until we return those regular items to editor. This combined with the existing behavior of not showing unimported items until they are calculated, would mean increased chance users won't see those items for the first few characters they typed.

Reply:
makes sense.  thanks.